staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go

```diff
@@ -154,9 +156,21 @@ var DefaultEndpointRestrictions = emptyEndpointRestrictions{}
 
 type emptyEndpointRestrictions struct{}
 
-func (emptyEndpointRestrictions) AllowsConversion(schema.GroupVersionKind) bool { return false }
-func (emptyEndpointRestrictions) AllowsServerVersion(string) bool               { return false }
-func (emptyEndpointRestrictions) AllowsStreamSchema(s string) bool              { return s == "watch" }
+func (emptyEndpointRestrictions) AllowsConversion(gvk schema.GroupVersionKind) bool {
+	if gvk.GroupVersion() == metav1alpha1.SchemeGroupVersion {
+		switch gvk.Kind {
+		case "Table", "PartialObjectMetadata", "PartialObjectMetadataList":
```

Comment:
hmm, I didn't expect to see specific types here

Reply:
actually, isn't the point of emptyEndpointRestrictions to avoid matching any specific type request?

Reply:
Yes

Reply:
I think the reason I add specific types here is same as here: (a temporary solution )  https://github.com/kubernetes/kubernetes/blob/fe13072443289a87ac2bf89fa7818f0ba8a5c64d/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L83, if we avoid matching any specific type, then the regression test like this in unit test here https://github.com/kubernetes/kubernetes/blob/46648a5eaf22bcfdb66825680ba5f4bdd74bbd4c/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate_test.go#L210 will be no meaning. since no matter how long and most specific your Accept header is, even they are correct form, like `application/json;as=PartialObjectMetadataList;v=v1alpha1;g=meta.k8s.io` the unit test will always pick the short one, since `acceptMediaTypeOptions` always return false

Reply:
@liggit do you have any suggestions ?

Reply:
The linked unit test should be selecting application/json.  In the absence of qualifiers, the first matching, valid accept header should be used

Reply:
The function you linked should only return true if the endpoint supports those conversions (all objects support partial object meta, only endpoints that have a TableConvertor should support the table object).

Reply:
The emptyEndpointRestrictions is used in both `NegotiateOutputSerializer` and `NegotiateOutputStreamSerializer`.
* `NegotiateOutputSerializer` is called in the discovery endpoints. They don't support Table Convertor. The types served don't have ObjectMeta at all, so the original code is more appropriate.
* `NegotiateOutputStreamSerializer` is called by the watch handler. Shouldn't the watch handler instead get the endpoint restrictions from the [scope](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L65)? @smarterclayton 

@shiywang As for the unit test, I think you can add a test testing `NegotiateOutputMediaType`, and add custom endpoint restrictions in your test matrix.

Reply:
Empty should not support anything.  Both restrictions should be provided by the scope, and the underlying type has to support the rules for partial metadata.  Only types that implement the `meta.Object` interface are candidates for PartialObjectMetadata.  Only storage implementations with ConvertToTable should support TableOutput.

Reply:
> @shiywang As for the unit test, I think you can add a test testing NegotiateOutputMediaType, and add custom endpoint restrictions in your test matrix.

Thanks @caesarxuchao I will add a custom endpoint restriction.

> The linked unit test should be selecting application/json. In the absence of qualifiers, the first matching, valid accept header should be used

why the first matching, left a question here: https://github.com/kubernetes/kubernetes/pull/50603/files#r153130192